### DISCUSSION TITLE
Better way to handle multiple custom middleware?

### DISCUSSION THREAD

**Discussion by jordantshaw** [fastapi/fastapi#6223] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class TimingMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        begin = time.time()
        response = await call_next(request)
        response_time = time.time() - begin
        response.headers["X-Response-Time"] = str(response_time)
        return response


class LoggingMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        logger.info({
            'message': 'request log',
            'request': request,
            'method': request.method,
            'url': request.url
        })

        response = await call_next(request)

        logger.info({
            'message': 'response log',
            'response': response,
            'status': response.status_code,
            'response_time': response.headers.get('X-Response-Time')
        })

        return response


class RequestContextMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        correlation_id_token = _correlation_id_ctx_var.set(request.headers.get('X-Correlation-ID', str(uuid4())))

        try:
            response = await call_next(request)
            response.headers['X-Correlation-ID'] = get_correlation_id()
        except Exception as err:
            raise err
        finally:
            _correlation_id_ctx_var.reset(correlation_id_token)

        return response



app.add_middleware(middleware.TimingMiddleware)
app.add_middleware(middleware.LoggingMiddleware)
app.add_middleware(middleware.RequestContextMiddleware)
```

### Description

I have a FastAPI application that uses multiple custom middleware classes.  Currently, I have created each middleware as its own class, and then add the middleware to the app with "add_middleware" method.

However, because of the nature of the middleware, the order in which the middleware get added is important.  Is there a better or cleaner way to be creating and registering these middleware?  Should I just be creating one main middleware class, that contains all the functionality, or is better to split them up like I have them?  I dont like the fact that the order in which each one is added to the app matters. 

### Operating System

macOS

### FastAPI Version

0.75.2

### Python Version

Python 3.9.12

**Comment by JarroVGIT** [comment]
I understand that the ordering of Middleware classes is not really something that is desired. However, from a maintainability and reusability point of view I think splitting them up is the right approach. 

When you would combine them in one big blob of logic, it would be very hard to (unit) test.

An exception to the above would be where the ordering of code prior `call_next()` should be different than the code after `call_next()`. For example, your code now does this:
```
request -> 
    (1)start timer ->
        (2)log request ->
            (3)set correlation id ->
                    respons ->
            (3)add correlation id to response header ->
        (2)log response ->
    (1)add time header ->
response ->
```
However, if you want to log the response with the timer header already set, you would need to change that around:
```
request -> 
    (1)start timer ->
        (2)log request ->
            (3)set correlation id ->
                    respons ->
            (3)add correlation id to response header ->
        (1)Ladd time header ->
    (2)log response->
response ->
```

In such a case, I can imagine that you combine where it makes sense. However, I would still recommend to make your code as maintainable and reusable as possible. The way you have structured it now would allow your colleagues to reuse bits and pieces rather than to untangle complicated ordering logic. 

Hope this helps!

**Comment by adriangb** [comment, marked as answer]
First off, I **highly recommend you do not use `BaseHTTPMiddleware`**: https://www.starlette.io/middleware/#basehttpmiddleware

Second, what I would do in your case is pre-compose this middlewares.
First off, I'm assuming this is some sort of shared library.
If this is a single app, I would suggest you add a comment explaining that the order matters please don't touch it, a test _if you really want to_ and otherwise go have a 🍺 , this is not something that would concern me in a code review.

Assuming this is some sort of library or you really want to enforce order, there's plenty of ways to do that without creating a single huge middleware.
Since none of your middleware have constructor parameters, this should be pretty straightforward.

## Option 1: export a list of middlewares

middleware.py
```python
from starlette.middleware import Middleware

my_pre_ordered_middlewares = [Middleware(TimingMiddleware), Middleware(LoggingMiddleware)]`

...

# in the application code
app = FastAPI(middleware=my_pre_ordered_middleware)
```

## Option 2: a hook

```python
def add_my_middleware(app: FastAPI) -> None:
    app.add_middleware(...)
    app.add_middleware(...)
```

## Option 3: a meta-middleware

```python
class MetaMiddleware:

    def __init__(self, app: FastAPI) -> None:
        self._app = TimingMiddleware(LoggingMiddleware(app))

   async def __call__(self, scope, receive, send):
        await self._app(scope, receive, send)
```

You could make this some sort of fancy factory, but I'd keep it simple.
